### ISSUE TITLE
[BUG] cython 3.0.0b1 stdlib.pxd qsort extern import breaks

### ISSUE THREAD

**Issue by mlguruz** [cython/cython#5267]
### Describe the bug

in stdlib.pxd, we have 
```
cdef extern from "<stdlib.h>" nogil:
    void qsort (void *array, size_t count, size_t size,
                int (*compare)(const void *, const void *))
```

in sklearn here: https://github.com/scikit-learn/scikit-learn/blob/e6b46675318950fd5138aaf29ba76f6ce2fd91b8/sklearn/tree/_splitter.pyx#L1091, this comparator is later pass to qsort here: https://github.com/scikit-learn/scikit-learn/blob/e6b46675318950fd5138aaf29ba76f6ce2fd91b8/sklearn/tree/_splitter.pyx#L1188

this gives error msg: 

`Cannot assign type 'int (const void *, const void *) except? -1 nogil' to 'int (*)(const void *, const void *) noexcept nogil'`

in the new cython 3.0.0b1 release.

I fixed it by doing the following:

```
ctypedef int (*qsortComparator_t)(const void*, const void*)

cdef extern from "<stdlib.h>" nogil:
    void qsort (void *array, size_t count, size_t size, qsortComparator_t comp)
```

...instead of importing from `stdlib.pxd`.

Since `stdlib.pxd` is native to Cython, maybe we could fix that so that downstream users don't have to custom extern import it? 

### Code to reproduce the behaviour:

```cython
# example code
```

### Environment

OS: Linux
Python version 3.8.3
Cython version 3.0.0b1

**Comment by da-woods** [comment]
Thanks for the report. Not sure I understand *exactly* what's gone wrong on first glance (it's to do with cdef functions being `except` by default, but don't know what's happening beyond that yet) but we'll definitely want to fix it

**Comment by matusvalo** [comment]
I tried to do replicate the issue but I am not able to:
```cython
from libc.stdlib cimport qsort

cdef void deleteme(void *array, size_t count, size_t size):
    qsort(array, count, size, compare_SIZE_t)
                  

cdef int compare_SIZE_t(const void* a, const void* b) noexcept nogil:
    return <int>((<int*>a)[0] - (<int*>b)[0])
```

This compiles successfully.

**Comment by scoder** [comment]
I'm also not able to reproduce the issue on my side. @mlguruz, are you sure that you tried it with the comparison function that had https://github.com/scikit-learn/scikit-learn/commit/dfe9e2e118bd40fa4cd3fce7e007bc7bf51c9033 applied?

**Comment by mlguruz** [comment]
> I'm also not able to reproduce the issue on my side. @mlguruz, are you sure that you tried it with the comparison function that had [scikit-learn/scikit-learn@dfe9e2e](https://github.com/scikit-learn/scikit-learn/commit/dfe9e2e118bd40fa4cd3fce7e007bc7bf51c9033) applied?

@scoder you're right - I wasn't using the sklearn's latest main branch. 

To make sure - I tested again using the main branch and default `qsort` from stdlib.pxd, and running below:

```
# cdef int compare_SIZE_t(const void* a, const void* b) nogil:                          # <- this fails
cdef int compare_SIZE_t(const void* a, const void* b) noexcept nogil:             # <- this works
    """Comparison function for sort."""
    return <int>((<SIZE_t*>a)[0] - (<SIZE_t*>b)[0])
```

Removing `noexcept` will trigger the compiling failure. Adding `noexcept` will compile fine. 

Is this the expected behavior?

**Comment by matusvalo** [comment]
> Is this the expected behavior?

Yes. Cython 3 changed the handling of exception in functions: https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#arithmetic-special-methods

**Comment by mlguruz** [comment]
> > Is this the expected behavior?
> 
> Yes. Cython 3 changed the handling of exception in functions: https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#arithmetic-special-methods

Thank you - this makes sense.
Seems like I should add the `legacy_implicit_noexcept` to a few places to make it backward compatible.
Other than that, please feel free to close this ticket. 
Thanks again!
